components/concurrency_manager/src/lib.rs

```diff
@@ -52,9 +104,100 @@ impl ConcurrencyManager {
 
     /// Updates max_ts with the given new_ts. It has no effect if
     /// max_ts >= new_ts or new_ts is TimeStamp::max().
-    pub fn update_max_ts(&self, new_ts: TimeStamp) {
-        if new_ts != TimeStamp::max() {
-            self.max_ts.fetch_max(new_ts.into_inner(), Ordering::SeqCst);
+    ///
+    /// To avoid invalid ts breaking the invariants, the new_ts should be
+    /// less than or equal to the max_ts_limit.
+    ///
+    /// # Returns
+    /// - Ok(()): If the update is successful or has no effect
+    /// - Err(limit): If new_ts is greater than the max_ts_limit, returns the
+    ///   current limit value
+    pub fn update_max_ts(
+        &self,
+        new_ts: TimeStamp,
+        source: impl slog::Value + Display,
```

Comment:
OK I tried them:
Using `format!` directly:
`time:   [14.652 ns 14.683 ns 14.716 ns]`
Using closure:
`time:   [14.887 ns 14.910 ns 14.933 ns]`
Using `format_args!`:
`time:   [4.1241 ns 4.1254 ns 4.1281 ns]`

I didn't manage to reproduce your result. If you've verified it could you also help test with `format_args!` and see the diff?

Reply:
Oh `format_args!` is good enough! Should this [`format!`](https://github.com/tikv/tikv/blob/742c826bccd95ca02ef273451a9e06c42e8f60fe/src/storage/txn/actions/prewrite.rs#L157) also be a `format_args!`?

```
update_max_ts_valid 16.154 ns
update_max_ts_valid_args 4.9605 ns
update_max_ts_invalid 19.849 ns
update_max_ts_invalid_args 8.2091 ns

update_max_ts_lazy_valid 3.8556 ns
update_max_ts_lazy_invalid 7.4589 ns
```

I test with this commit. https://github.com/you06/tikv/commit/1074553e313b6060b1d337fc129d46802db7f218

Reply:
OK I see. I've changed it so that either strings or closures can be accepted.